### ISSUE TITLE
Cursor jumps to end of input fields

### ISSUE THREAD

**Issue by DAreRodz** [frontity/frontity#236]
This bug was first reported here: https://community.frontity.org/t/cursor-jumping-on-controlled-input/789

**_Bug report_**

- [x] I've searched for existing issues
- [x] I've checked documentation: https://docs.frontity.org
- [x] I've checked community forum: https://community.frontity.org

**Describe the expected behavior**

While tipying in an `<input>` component, the cursor must remain in place.

**Describe the observed behavior**

The cursor jumps to the end every time the value changes.

**Describe the steps involved to reproduce the problem**

Just add an `<input>` to your theme and test it.

**Comment by DAreRodz** [comment]
It seems a problem that many people have already had. Just to keep a reference, I leave here a couple of links:

https://github.com/facebook/react/issues/955#issuecomment-469352730
https://github.com/apollographql/apollo-client/issues/3338#issuecomment-552190902

**Comment by luisherranz** [comment]
Yeah, that problem comes from the monkey patching of the native events done by react-easy-state to batch events under React's `unsafe_batchedUpdates`:
https://github.com/solkimicreb/react-easy-state/issues/92

I already offered my help to solve it in the original library and even found a solution that we could use in our own fork:
https://codesandbox.io/s/react-easy-state-batch-bug-async-scheduler-6q2dt

I was waiting to hear from @solkimicreb first, to understand why he monkey patched instead of using async queues, but it seems like [he is still missing](https://github.com/solkimicreb/react-easy-state/issues/95).

If this is urgent we can implement my solution. Why did you opened this issue in the first place? Did some user reported the problem?

**Comment by DAreRodz** [comment]
Yes, some user did it in the community. It's in the issue description. ☝️

**Comment by luisherranz** [comment]
Ups, sorry.

**Comment by nitelite** [comment]
> I already offered my help to solve it in the original library and even found a solution that we could use in our own fork:
> https://codesandbox.io/s/react-easy-state-batch-bug-async-scheduler-6q2dt

@luisherranz Do you happen to have this change in a react-easy-state fork? I would love to be able to get this solved in my project.

**Comment by luisherranz** [comment]
I'm sorry @nitelite but our fork of `react-easy-state` doesn't have the same API and we are about to change it even more... besides, we haven't solved the bug yet.

But I added some instructions about how to do the change using `patch-package` in this issue of `react-easy-state`.

**Comment by luisherranz** [comment]
It looks like there is some movement on the https://github.com/RisingStack/react-easy-state repo and it's been moved to the RisingStack org. Let's see if they want to fix this issue.

**Comment by luisherranz** [comment]
Miklos is back and they have released a fix for this bug: https://github.com/RisingStack/react-easy-state/pull/113

**Comment by luisherranz** [comment]
By the way, the PR is huge because they included lots of format changes. Look for the `view.js`, `schedule.js` and `queue.js` files.

**Comment by luisherranz** [comment]
After some debugging, I have discovered that the current `scheduler` is the source of the problem with React DevTools described in this topic: https://community.frontity.org/t/invalid-renderer-id-1/1148/10

So once we solve this, that problem will also go away. I'm going to promote this issue to `priority: high` for that reason.

**Comment by luisherranz** [comment]
Also relevant: https://github.com/RisingStack/react-easy-state/pull/144

**Comment by solkimicreb** [comment]
Hi! I am the author of react-easy-state and I got an email about this issue. Just wanted to let you know that next release (with the cursor jumping fix) is coming early next week.

**Comment by luisherranz** [comment]
Yeah. Awesome, thanks Miklos :)

For your comment in the last PR (https://github.com/RisingStack/react-easy-state/pull/144):
> // React batches setStates in its event listeners anyways
> // so this commenting this part out is not a huge issue.

I guess your preference is to just disable batching in event handlers.

I comment this because I used `@nx-js/queue-util` to do a quick test and it seemed like that approach solved the batching problem without disabling event handlers: https://codesandbox.io/s/react-easy-state-batch-bug-async-scheduler-6q2dt

If I'm not mistaken, you moved away from this async queue approach in older versions of `react-easy-state` to the batching approach in the last version. May I know why? I'd love to know what made you take that decision.

Apart from that, I also did an experiment with React Concurrent and batching won't be needed anymore once they release it, so that's nice to know: https://github.com/RisingStack/react-easy-state/issues/92#issuecomment-546686630 🙂

**Comment by solkimicreb** [comment]
> Apart from that, I also did an experiment with React Concurrent and batching won't be needed anymore once they release it, so that's nice to know.

Yep, I can't wait for that 🙂 

> If I'm not mistaken, you moved away from this async queue approach in older versions of react-easy-state to the batching approach in the last version. May I know why? I'd love to know what made you take that decision.

It moved `setState`s that were inside React event handlers to outside of them (in a new task) which caused a huge performance hit in some cases. It is probably related to some internal React stuff that I never figured out.

Edit: I guess it is related to this:

"This is crucial for good performance in large applications because if Child and Parent each call setState() when handling a click event, you don't want to re-render the Child twice."

To me this means that React is not just batching setStates in event handlers but it is also deduping renders triggered by other means (like props).

Edit2: `unstable_batchedUpdates` was not yet around when I removed the async batching. Maybe combining the two would be worth a shot. (Until React 17 is released).

Edit3: (Sorry, memories are coming back 😄) Async batching really messes up tests for most people since they expect to be able to do sync expectations (as with normal current React). This is actually a biggie.

**Comment by luisherranz** [comment]
> Async batching really messes up tests for most people since they expect to be able to do sync expectations

Absolutely true.

Actually, that is the first thing I stumbled upon when doing the [test of `react-easy-state` with Concurrent mode](https://codesandbox.io/s/react-easy-state-batch-bug-async-scheduler-6q2dt) and for that reason, I did those weird tests in codesandbox. But that is something that will have to be fixed by both React and React testing libraries once Concurrent mode is out.

Anyway, many thanks for stopping by Miklos :)
